apps/vmq_server/src/vmq_mqtt10_fsm.erl

```diff
@@ -0,0 +1,1351 @@
+    {error, not_allowed} when ?IS_PROTO_4(Proto) ->
+      %% we have to close connection for 3.1.1
+      _ = vmq_metrics:incr_mqtt_error_auth_publish(),
+      {error, not_allowed};
+    {error, _Reason} ->
+      %% can't publish due to overload or netsplit
+      _ = vmq_metrics:incr_mqtt_error_publish(),
+      []
+  end.
+
+-spec dispatch_publish_qos1(msg_id(), msg(), state()) ->
+  list()
+  | {list(), session_ctrl()}
+  | {error, not_allowed}.
+dispatch_publish_qos1(MessageId, Msg, State) ->
+  #state{username=User, subscriber_id=SubscriberId, proto_ver=Proto,
+    cap_settings=CAPSettings, reg_view=RegView} = State,
+  case publish(CAPSettings, RegView, User, SubscriberId, Msg) of
+    {ok, _, SessCtrl} ->
+      _ = vmq_metrics:incr_mqtt_puback_sent(),
+      {[#mqtt_puback{message_id=MessageId}], SessCtrl};
+    {error, not_allowed} when ?IS_PROTO_4(Proto) ->
+      %% we have to close connection for 3.1.1
+      _ = vmq_metrics:incr_mqtt_error_auth_publish(),
+      {error, not_allowed};
+    {error, not_allowed} ->
+      %% we pretend as everything is ok for 3.1 and Bridge
+      _ = vmq_metrics:incr_mqtt_error_auth_publish(),
+      _ = vmq_metrics:incr_mqtt_puback_sent(),
+      [#mqtt_puback{message_id=MessageId}];
+    {error, _Reason} ->
+      %% can't publish due to overload or netsplit
+      _ = vmq_metrics:incr_mqtt_error_publish(),
+      []
+  end.
+
+-spec dispatch_publish_qos2(msg_id(), msg(), state()) ->
+  list()
+  | {state(), list(), session_ctrl()}
+  | {error, not_allowed}.
+dispatch_publish_qos2(MessageId, Msg, State) ->
+  #state{username=User, subscriber_id=SubscriberId, proto_ver=Proto,
+    cap_settings=CAPSettings, reg_view=RegView, waiting_acks=WAcks} = State,
+  case maps:get({qos2, MessageId}, WAcks, not_found) of
+    not_found ->
+      case publish(CAPSettings, RegView, User, SubscriberId, Msg) of
+        {ok, _, SessCtrl} ->
+          Frame = #mqtt_pubrec{message_id=MessageId},
+          _ = vmq_metrics:incr_mqtt_pubrec_sent(),
+          {State#state{
+            waiting_acks=maps:put({qos2, MessageId}, Frame, WAcks)},
+            [Frame], SessCtrl};
+        {error, not_allowed} when ?IS_PROTO_4(Proto) ->
+          %% we have to close connection for 3.1.1
+          _ = vmq_metrics:incr_mqtt_error_auth_publish(),
+          {error, not_allowed};
+        {error, not_allowed} ->
+          %% we pretend as everything is ok for 3.1 and Bridge
+          _ = vmq_metrics:incr_mqtt_error_auth_publish(),
+          _ = vmq_metrics:incr_mqtt_pubrec_sent(),
+          Frame = #mqtt_pubrec{message_id=MessageId},
+          [Frame];
+        {error, _Reason} ->
+          %% can't publish due to overload or netsplit
+          _ = vmq_metrics:incr_mqtt_error_publish(),
+          []
+      end;
+    _Frame ->
+      Frame = #mqtt_pubrec{message_id=MessageId},
+      [Frame]
+  end.
+
+%%ToDo: Do not transfer qos0 acks to offline store. Also remove pending puback frames.
```

Comment:
PubAck frames means qos 0 puback frames that need to be sent to the Publisher. This does not require handling because PubAcks are never stored in the WAcks in the first place.